storage/innobase/fts/fts0fts.cc

```diff
@@ -6206,3 +5994,464 @@ fts_init_index(
+                                          m_tuple, &mtr);
+    btr_pcur_store_position(&pcur, &mtr);
+    mtr.commit();
+
+    if (err && !handle_error(&err))
+      goto err_exit;
+
+    mtr.start();
+    if (pcur.restore_position(BTR_MODIFY_LEAF, &mtr) ==
+          btr_pcur_t::CORRUPTED)
+    {
+      err= DB_CORRUPTION;
+      goto func_exit;
+    }
+  }
+func_exit:
+  mtr.commit();
+err_exit:
+  btr_pcur_close(&pcur);
+  return err;
+}
+
+dberr_t FTSQueryExecutor::search_tuple(dict_table_t *table,
+                                       read_record *func, void *arg)
+{
+  dict_index_t *index= dict_table_get_first_index(table);
+  mtr_t mtr;
+  m_search_pcur= static_cast<btr_pcur_t*>(mem_heap_zalloc(m_heap,
+                                                          sizeof(btr_pcur_t)));
+  rec_offs *offsets= nullptr;
+  buf_block_t *block= nullptr;
+  rec_t *rec= nullptr;
+  mtr.start();
+  mtr.set_named_space(table->space);
+
+  m_search_pcur->btr_cur.page_cur.index= index;
+  dberr_t err= btr_pcur_open_on_user_rec(m_tuple, BTR_SEARCH_LEAF,
+                                         m_search_pcur, &mtr);
+  if (err != DB_SUCCESS)
+    goto func_exit;
+
+  while ((rec= btr_pcur_get_rec(m_search_pcur)))
+  {
+    while (page_rec_is_infimum(rec) || page_rec_is_supremum(rec)
+           || rec_get_deleted_flag(rec, dict_table_is_comp(table)))
+    {
+      if (!btr_pcur_move_to_next(m_search_pcur, &mtr))
+      {
+        err= DB_END_OF_INDEX;
+        goto func_exit;
+      }
+      rec= btr_pcur_get_rec(m_search_pcur);
+    }
+
+    /* Find offsets */
+    offsets= rec_get_offsets(rec, index, nullptr,
+                             index->n_core_fields, ULINT_UNDEFINED,
+                             &m_heap);
+
+    if (cmp_dtuple_rec(m_tuple, rec, index, offsets) != 0)
+    {
+      err= DB_RECORD_NOT_FOUND;
+      goto func_exit;
+    }
+
+    block= btr_pcur_get_block(m_search_pcur);
+    err= lock_clust_rec_read_check_and_lock(0, block, rec, index, offsets,
+                                            LOCK_S, LOCK_ORDINARY, m_thr);
+    if (err == DB_SUCCESS_LOCKED_REC) err= DB_SUCCESS;
+
+    if (err)
+    {
+      /* Store the position */
+      btr_pcur_store_position(m_search_pcur, &mtr);
+      mtr.commit();
+
+      if (!handle_error(&err)) goto err_exit;
+
+      /* Lock wait */
+      mtr.start();
+      if (m_search_pcur->restore_position(BTR_SEARCH_LEAF, &mtr)
+          == btr_pcur_t::CORRUPTED)
+      {
+        err= DB_CORRUPTION;
+        goto func_exit;
+      }
+    }
+    else
+    {
+      if (func)
+       (*func)(rec, offsets, arg);
+
+      if (!btr_pcur_move_to_next(m_search_pcur, &mtr))
+      {
+        err= DB_END_OF_INDEX;
+        goto func_exit;
+      }
+    }
+  }
+func_exit:
+  mtr.commit();
+err_exit:
+  btr_pcur_close(m_search_pcur);
+  return err;
+}
+
+dberr_t FTSQueryExecutor::delete_all(dict_table_t *table)
+{
+  dict_index_t *index= dict_table_get_first_index(table);
```

Comment:
Shouldn’t we assert that there only is one index in the table? Otherwise this function would corrupt any secondary indexes.